compiler/rustc_infer/src/traits/engine.rs

```diff
@@ -17,6 +17,8 @@ use crate::traits::Obligation;
 pub enum ScrubbedTraitError<'tcx> {
     /// A real error. This goal definitely does not hold.
     TrueError,
+    // A select error with pending obligations
+    Select(PredicateObligations<'tcx>),
```

Comment:
I understand that it's not really documented, but this is antithetical to the design of `ScrubbedTraitError`. That type should be very lightweight.

Reply:
Totally understand your concern. I tried this mainly as a quick hack. I did this because through my ad hoc debugging, I uncovered that `select_where_possible` contains the pending constraints that I needed to display within the error message. But since the returned details are erased via `ScrubbedTraitError`, I tried to pass on the pending obligations so that I can show it inside the error message.

I'm not sure what is the best way to workaround this, but at least with this quick hack, I'm able to use this fork of the Rust compiler to significantly simplify debugging for my projects. So regardless of which approach is taken, I probably do need to extract the pending obligations in one way or another and then surface it inside my error messages.

Btw, I have also considered alternative approaches, such as writing a compiler plugin to extract the pending obligations for my specific use cases. However, as far as I looked, I couldn't find ways to implement compiler plugins that help recover or better report compile errors. So I do hope that I can find some ways to fix this upstream in Rust, so that I don't need to instruct my users to use my custom fork of Rust compiler to use my project.